File: dotnet/Challenges/FunctionalChallenges/BitsAndBytes.cs
namespace Challenges;

public static class BitsAndBytes
{
    // Original: int bitcount(unsigned x), "The C Programming Language", page 50
    public static int BitCount(uint x) =>
        x == 0 ? 0 : BitCount(x >> 1) + (int)(x & 1);

    // In the next three methods I use [..size] instead of slicing with Span<T> for more compact code
    // But .Slice(0, size) is more probably more readable
    
    public static void ByteCopy(byte[] src, byte[] dst, int size) =>
        ((Span<byte>)src)[..Min([size, src.Length, dst.Length])].CopyTo(dst);

    public static void ByteZero(byte[] a, int size) =>
        ((Span<byte>)a)[..Min([a.Length, size])].Clear();

    public static void ByteSet(byte[] a, byte value, int size) =>
        ((Span<byte>)a)[..Min([a.Length, size])].Fill(value);
    
    private static int Min(int[] numbers) => numbers.MinBy(n => n);
}

File: typescript/challenges/src/bits-and-bytes.ts
// Original: int bitcount(unsigned x), "The C Programming Language", page 50

export const bitCount = (x: number): number => _bitCount(x, 0);

const _bitCount = (x: number, iterations: number): number =>
    x == 0 || iterations == 32
        ? 0
        : _bitCount(x >> 1, ++iterations) + (x & 1);

export const byteCopy = (src: Uint8Array, dst: Uint8Array, size: number) : void => src
    .slice(0, Math.min(size, Math.min(src.length, dst.length)))
    .forEach((value, i) => dst[i] = value);

export function byteZero(a: Uint8Array, size: number) : void {
    a.fill(0, 0, Math.min(size, a.length));
}

export function byteSet(a: Uint8Array, value: number, size: number) : void {
    a.fill(value, 0, Math.min(size, a.length));
}